frontend/src/pages/AllJobsList.test.tsx

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 Arrikto Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from 'react';
+import { AllJobsList } from './AllJobsList';
+import { PageProps } from './Page';
+import { RoutePage } from '../components/Router';
+import { shallow, ShallowWrapper } from 'enzyme';
+import { ButtonKeys } from '../lib/Buttons';
+import TestUtils from '../TestUtils';
+
+describe('AllJobsList', () => {
+  const updateBannerSpy = jest.fn();
+  let _toolbarProps: any = {};
+  const updateToolbarSpy = jest.fn(toolbarProps => (_toolbarProps = toolbarProps));
+  const historyPushSpy = jest.fn();
+  const props: PageProps = {
+    history: { push: historyPushSpy } as any,
+    location: '' as any,
+    match: '' as any,
+    toolbarProps: _toolbarProps,
+    updateBanner: updateBannerSpy,
+    updateDialog: jest.fn(),
+    updateSnackbar: jest.fn(),
+    updateToolbar: updateToolbarSpy,
+  };
+  let tree: ShallowWrapper;
+
+  function shallowMountComponent(
+    propsPatch: Partial<PageProps & { namespace?: string }> = {},
+  ): void {
+    tree = shallow(<AllJobsList {...props} {...propsPatch} />);
+    // Necessary since the component calls updateToolbar with the toolbar props,
+    // then expects to get them back in props
+    tree.setProps({ toolbarProps: _toolbarProps });
+    updateToolbarSpy.mockClear();
+  }
+
+  beforeEach(() => {
+    updateBannerSpy.mockClear();
+    updateToolbarSpy.mockClear();
+    historyPushSpy.mockClear();
+  });
+
+  afterEach(() => tree.unmount());
+
+  it('renders all jobs', () => {
+    shallowMountComponent();
+    expect(tree).toMatchSnapshot();
+  });
+
+  it('lists all jobs in namespace', () => {
+    shallowMountComponent({ namespace: 'test-ns' });
+    expect(tree.find('JobList').prop('namespaceMask')).toEqual('test-ns');
+  });
+
+  it('removes error banner on unmount', () => {
+    shallowMountComponent();
+    tree.unmount();
+    expect(updateBannerSpy).toHaveBeenCalledWith({});
+  });
+
+  // it('refreshes the job list when refresh button is clicked', async () => {
```

Comment:
In shallow rendering mode, all child components are mocked, so you cannot call refresh there.

You can follow best practice: https://github.com/kubeflow/pipelines/issues/5118#issuecomment-776177682, to use react-testing-library instead. Render everything and query&click the refresh button instead of calling the function.

Reply:
@Bobgy thanks this make a lot of sense. So I tried to play around with `render` but I think I am missing something...See here https://github.com/kubeflow/pipelines/pull/5131/commits/da09a9596c8cc5d9938462e2ba727deeb55727bd I tried to follow a similar approach as here https://github.com/kubeflow/pipelines/blob/1fee4054a77d39877d997ec94374201188fa87d0/frontend/src/pages/RunDetails.test.tsx#L1700 but in the AllJobsList case I can't get any child element, I always get undefined. What  am I missing?

Reply:
Also, I noticed that even in the `RunDetails` case, I cannot get by any means one of the toolbar buttons from `tree`, are they even rendered?

Reply:
Ohhh, you are right. The root page is not rendered. Let me think about it.

Reply:
I think we can defer the test to later PRs.

A proper solution should be building a test page wrapper (or maybe use the real page wrapper) that actually render these buttons.